### PR TITLE
chore: bump to latest alloy and revm

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Bump alloy. Fixes https://github.com/foundry-rs/foundry/issues/7932

## Solution

Redo RPC calls

### CODE REVIEW

**Comment by yash-atreya** [comment]
Supersedes #7933

File: Cargo.lock [hunk 1]
```diff
@@ -6379,7 +6395,17 @@ version = "4.0.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "d322f2730cd300e99d271a1704a2dfb8973d832428f5aa282aaa40e2473b5eec"
 dependencies = [
- "revm-primitives",
+ "revm-primitives 3.1.1",
```
Comment (DaniPopes) [inline Cargo.lock:6398]: revm must also be bumped.

Comment (yash-atreya) [reply Cargo.lock:6398]: Yes on it

File: crates/evm/core/src/ic.rs [hunk 2]
```diff
@@ -43,7 +43,7 @@ impl IcPcMap {
 }
 
 fn make_map<const PC_FIRST: bool>(spec: SpecId, code: &[u8]) -> FxHashMap<usize, usize> {
-    let opcode_infos = spec_opcode_gas(spec);
+    let opcode_infos = spec_opcode_gas(spec); // TODO(yash): What's the replacement for spec_opcode_gas in revm 9?
```
Comment (DaniPopes) [inline crates/evm/core/src/ic.rs:46]: in this specific case you can replace it with opcode infos and get the immediate size. We're just iterating over bytecode here

File: crates/cast/bin/cmd/mktx.rs [hunk 3]
```diff
@@ -90,8 +90,9 @@ impl MakeTxArgs {
         tx::validate_from_address(eth.wallet.from, from)?;
 
         if resend {
-            tx.nonce =
-                Some(U64::from(provider.get_transaction_count(from, BlockId::latest()).await?));
+            tx.nonce = Some(U64::from(
+                provider.get_transaction_count(from).block_id(BlockId::latest()).await?,
```
Comment (mattsse) [inline crates/cast/bin/cmd/mktx.rs:94]: the latest call is not required because latest is the default

**Review by mattsse — CHANGES_REQUESTED** [review]
early comment

File: crates/evm/core/src/backend/mod.rs [hunk 4]
```diff
@@ -760,14 +760,20 @@ impl Backend {
...
-            }
-            TransactTo::Create(CreateScheme::Create2 { salt }) => {
-                let code_hash = B256::from_slice(keccak256(&env.tx.data).as_slice());
-                env.tx.caller.create2(B256::from(salt), code_hash)
+            TransactTo::Call(to) => {
+                if to != DEFAULT_CREATE2_DEPLOYER {
+                    to
+                } else {
+                    let code_hash = B256::from_slice(keccak256(&env.tx.data).as_slice());
+                    env.tx.caller.create2(B256::from(salt), code_hash) // TODO(yash): Get CREATE2
```
[the remark this answers is not in the document]
Comment (yash-atreya) [reply crates/evm/core/src/backend/mod.rs:769]: Am I missing something here?

File: crates/cast/bin/cmd/mktx.rs (hunk 3 above)
Comment (DaniPopes) [reply crates/cast/bin/cmd/mktx.rs:94]: ues pls remove everywhere

File: crates/evm/core/src/backend/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/evm/core/src/backend/mod.rs:769]: I believe we can remove this, there is no create2 transaction, this was a small oversight in revm

File: crates/cast/bin/main.rs [hunk 5]
```diff
@@ -351,7 +351,8 @@ async fn main() -> Result<()> {
             let provider = utils::get_provider(&config)?;
             let address = address.resolve(&provider).await?;
             let value = provider
-                .get_proof(address, slots.into_iter().collect(), block.unwrap_or(BlockId::latest()))
+                .get_proof(address, slots.into_iter().collect())
+                .block_id(block.unwrap_or(BlockId::latest()))
```
Comment (DaniPopes) [inline crates/cast/bin/main.rs:355]: unwrap_or_default instead of unwrap_or(latest)

everywhere else too

**Review by DaniPopes — CHANGES_REQUESTED** [review]

**Comment by yash-atreya** [comment]
@DaniPopes @mattsse ptal

Some critical commits where I've addressed the breaking revm changes:

https://github.com/foundry-rs/foundry/pull/7934/commits/b84cf360dcd1718f667d01ade4b395745ac39902

https://github.com/foundry-rs/foundry/pull/7934/commits/db6182780d64dd2389c46d47e56fa8227e0575ec

https://github.com/foundry-rs/foundry/pull/7934/commits/b536ce030cf0bbb2a47fab196fbd4441d78c0880

https://github.com/foundry-rs/foundry/pull/7934/commits/559ec177b40d8f12c1fed1ef90934d5c93b3f45e

Mostly mapped `call.contract` to `call.target_address`

**Comment by yash-atreya** [comment]
> @DaniPopes @mattsse ptal
> 
> Some critical commits where I've addressed the breaking revm changes:
> 
> [b84cf36](https://github.com/foundry-rs/foundry/commit/b84cf360dcd1718f667d01ade4b395745ac39902)
> 
> [db61827](https://github.com/foundry-rs/foundry/commit/db6182780d64dd2389c46d47e56fa8227e0575ec)
> 
> [b536ce0](https://github.com/foundry-rs/foundry/commit/b536ce030cf0bbb2a47fab196fbd4441d78c0880)
> 
> [559ec17](https://github.com/foundry-rs/foundry/commit/559ec177b40d8f12c1fed1ef90934d5c93b3f45e)
> 
> Mostly mapped `call.contract` to `call.target_address`

Yeah CI is failing. Related to these commits. I'll fix.

File: crates/evm/core/src/ic.rs [hunk 6]
```diff
@@ -56,10 +55,10 @@ fn make_map<const PC_FIRST: bool>(spec: SpecId, code: &[u8]) -> FxHashMap<usize,
             map.insert(ic, pc);
         }
 
-        let op = code[pc];
-        if opcode_infos[op as usize].is_push() {
+        let op = unsafe { OpCode::new_unchecked(code[pc]) };
+        if op.is_push() {
```
Comment (DaniPopes) [inline crates/evm/core/src/ic.rs:59]: same

File: crates/evm/fuzz/src/strategies/state.rs [hunk 7]
```diff
@@ -7,8 +7,11 @@ use indexmap::IndexSet;
 use parking_lot::{lock_api::RwLockReadGuard, RawRwLock, RwLock};
 use revm::{
     db::{CacheDB, DatabaseRef, DbAccount},
-    interpreter::opcode::{self, spec_opcode_gas},
-    primitives::{AccountInfo, SpecId},
+    interpreter::{
+        opcode::{self},
```
Comment (DaniPopes) [inline crates/evm/fuzz/src/strategies/state.rs:11]:
```suggestion
        opcode,
```

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/evm/evm/src/inspectors/mod.rs [hunk 8]
```diff
@@ -18,3 +20,29 @@ pub use logs::LogCollector;
 
 mod stack;
 pub use stack::{InspectorData, InspectorStack, InspectorStackBuilder};
+
+use alloy_eips::eip2930::AccessList;
+use alloy_primitives::Address;
+/// To be used in Anvil
+pub struct AnvilAccessListInspector {
```
Comment (mattsse) [inline crates/evm/evm/src/inspectors/mod.rs:27]: please remove, we already had this

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by yash-atreya** [comment]
@DaniPopes how can I calc the `gasMemoryUsed` for a call now that `gas.memory()` has been deprecated in revm. The interpreter result doesn't expose `SharedMemory` which would give us access to `current_expansion_cost()` or the buffer len

https://github.com/foundry-rs/foundry/blob/860491e5ba37ffb1c7852f6478febb5c28837c8c/crates/cheatcodes/src/inspector.rs#L1112-L1119

**Comment by yash-atreya** [comment]
> @DaniPopes how can I calc the `gasMemoryUsed` for a call now that `gas.memory()` has been deprecated in revm. The interpreter result doesn't expose `SharedMemory` which would give us access to `current_expansion_cost()` or the buffer len
> 
> https://github.com/foundry-rs/foundry/blob/860491e5ba37ffb1c7852f6478febb5c28837c8c/crates/cheatcodes/src/inspector.rs#L1112-L1119

This is the last bit, it would solve failing CI tests

File: crates/evm/evm/src/inspectors/stack.rs [hunk 9]
```diff
@@ -551,19 +551,19 @@ impl InspectorStack {
         let (result, address, output) = match res.result {
             ExecutionResult::Success { reason, gas_used, gas_refunded, logs: _, output } => {
                 gas.set_refund(gas_refunded as i64);
-                gas.record_cost(gas_used);
+                let _record_gas = gas.record_cost(gas_used);
```
Comment (mattsse) [inline crates/evm/evm/src/inspectors/stack.rs:554]: can we remove the unused var?

**Review by mattsse — APPROVED** [review]
only 1 q re unused vars

File: crates/evm/evm/src/inspectors/stack.rs (hunk 9 above)
Comment (yash-atreya) [reply crates/evm/evm/src/inspectors/stack.rs:554]: Added this due to clippy, can remove or make it `let _ = gas.record_cost()`.

**Comment by yash-atreya** [comment]
> @DaniPopes how can I calc the `gasMemoryUsed` for a call now that `gas.memory()` has been deprecated in revm. The interpreter result doesn't expose `SharedMemory` which would give us access to `current_expansion_cost()` or the buffer len
> 
> https://github.com/foundry-rs/foundry/blob/860491e5ba37ffb1c7852f6478febb5c28837c8c/crates/cheatcodes/src/inspector.rs#L1112-L1119

Bump here @mattsse @DaniPopes

**Review by mattsse — CHANGES_REQUESTED** [review]
CI failures, looks like this is related to the memory recording

**Comment by yash-atreya** [comment]
> CI failures, looks like this is related to the memory recording

Yes, `gas.memory()` was deprecated in revm. The alternative is to use `current_expansion_cost` from `SharedMemory` but that is not exposed by the interpreter result.

**Comment by mattsse** [comment]
can we use the memory range here somehow and calc this manually @DaniPopes ?

not exactly sure what this previously tracked

**Comment by yash-atreya** [comment]
> can we use the memory range here somehow and calc this manually @DaniPopes ?
> 
> 
> 
> not exactly sure what this previously tracked

We need `buffer.len()` to calculate it manually and that's also in `SharedMemory`

File: crates/evm/evm/src/inspectors/debugger.rs [hunk 10]
```diff
@@ -51,11 +50,14 @@ impl<DB: DatabaseExt> Inspector<DB> for Debugger {
         let op = interp.current_opcode();
 
         // Get opcode information
-        let opcode_infos = spec_opcode_gas(ecx.spec_id());
-        let opcode_info = &opcode_infos[op as usize];
-
+        let _opcode_info = OPCODE_INFO_JUMPTABLE[op as usize]
+            .ok_or_else(|| eyre::eyre!("Invalid opcode: {}, Not found in jump table", op));
```
Comment (DaniPopes) [inline crates/evm/evm/src/inspectors/debugger.rs:54]: Let's remove these jumptable calls if we're not going to use the result. In EOF, more opcodes will have immediate values which we need to skip, but the entire iterator will look different anyway, so the manual check for push opcode is fine.

File: crates/evm/core/src/ic.rs [hunk 11]
```diff
@@ -56,10 +55,11 @@ fn make_map<const PC_FIRST: bool>(spec: SpecId, code: &[u8]) -> FxHashMap<usize,
             map.insert(ic, pc);
         }
 
-        let op = code[pc];
-        if opcode_infos[op as usize].is_push() {
+        let _op_info = OPCODE_INFO_JUMPTABLE[code[pc] as usize]
+            .ok_or_else(|| eyre::eyre!("Invalid opcode: {}, Not found in jump table", code[pc]));
+        if code[pc] >= PUSH1 && code[pc] <= PUSH32 {
```
Comment (DaniPopes) [inline crates/evm/core/src/ic.rs:60]: please use the same as in the other checks, so .contains

File: crates/evm/evm/src/inspectors/stack.rs (hunk 9 above)
Comment (DaniPopes) [reply crates/evm/evm/src/inspectors/stack.rs:554]: This is fine because it's a new Gas instance, I added must_use for use within EVM instructions.

**Review by DaniPopes — APPROVED** [review]
lgtm, nits, and as discussed we can make the lastCallGas memory gas 0

**Review by mattsse — APPROVED** [review]

**Comment by mattsse** [comment]
needs clippy

File: crates/cheatcodes/src/inspector.rs [hunk 12]
```diff
@@ -1119,8 +1115,8 @@ impl<DB: DatabaseExt> Inspector<DB> for Cheatcodes {
             gasLimit: gas.limit(),
             // The total gas used.
             gasTotalUsed: gas.spent(),
-            // The amount of gas used for memory expansion.
-            gasMemoryUsed: gas.memory(),
+            // DEPRECATED: The amount of gas used for memory expansion. Ref: https://github.com/foundry-rs/foundry/pull/7934#pullrequestreview-2069236939
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:1118]: pls remove these comments and move to doc comments on the cheatcodes gas struct

**Review by DaniPopes — APPROVED** [review]

**Comment by mattsse** [comment]
this test appears to be flaky testStorageAccessDelegateCall on multiple platforms, unsure if related to this pr

**Comment by DaniPopes** [comment]
likely some incorrect address migration in cheatcodes

**Comment by yash-atreya** [comment]
> likely some incorrect address migration in cheatcodes

Yeah I'll investigate

File: crates/cheatcodes/src/inspector.rs [hunk 13]
```diff
@@ -997,18 +993,19 @@ impl<DB: DatabaseExt> Inspector<DB> for Cheatcodes {
             // RecordedAccountAccess (and all subsequent RecordedAccountAccesses) will be
             // updated with the revert status of this call, since the EVM does not mark accounts
             // as "warm" if the call from which they were accessed is reverted
+
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:996]:
```suggestion
```

File: testdata/lib/forge-std
Comment (DaniPopes) [inline testdata/lib/forge-std:1]: please remove

**Comment by DaniPopes** [comment]
> 
this test appears to be flaky testStorageAccessDelegateCall on multiple platforms, unsure if related to this pr


@mattsse it's not flaky if it actually fails :D

**Review by DaniPopes — APPROVED** [review]

**Comment by mattsse** [comment]
great, just resolved 1 smol conflict

**Pull request merged**